### PR TITLE
[julia]: Update build script to be compatible with new BB

### PR DESCRIPTION
**Pull request by jstrube**
Not sure about the placement of `preferred_gcc_version`. It's not documented, so I guessed after looking at the source.

### CODE REVIEW

File: J/Julia/build_tarballs.jl [hunk 1]
```diff
@@ -72,32 +72,23 @@ make ${BUILD_FLAGS[@]} install
...
 platforms = [
-    Linux(:x86_64; libc=:glibc, compiler_abi=CompilerABI(:gcc7)),
-    Linux(:i686; libc=:glibc, compiler_abi=CompilerABI(:gcc7)),
-    Windows(:x86_64; compiler_abi=CompilerABI(:gcc7)),
-    Windows(:i686; compiler_abi=CompilerABI(:gcc7)),
+    Linux(:x86_64; libc=:glibc),
+    Linux(:i686; libc=:glibc),
+    Windows(:x86_64;),
+    Windows(:i686;),
 ]
```
Comment (ahumenberger) [inline J/Julia/build_tarballs.jl]: Why not ` platforms = supported_platforms()`?

**Comment by ahumenberger** [comment]
The problem seems to be that Julia cannot dlopen another version of libjulia in https://github.com/JuliaPackaging/BinaryBuilder.jl/blob/a8132866f9072508265d79f98d735885dd75b5b7/src/Auditor.jl#L114

**Comment by staticfloat** [comment]
Unfortunately, Julia cannot be cross-compiled, because the bootstrap portion of the build doesn't support cross-compilation.  It is possible to do it with WINE, but we don't have that setup in BB yet.  :(

**Comment by ViralBShah** [comment]
Can we make BB serve the regular Julia binaries (without trying to compile them)?

**Comment by Gnimuc** [comment]
Yes, I think we can do what [HDF5_jll](https://github.com/JuliaPackaging/Yggdrasil/blob/master/H/HDF5/build_tarballs.jl#L8-L25) did here as a temporary workaround.

**Comment by jstrube** [comment]
I understand how to download the tarballs for Linux and FreeBSD and extract them into the right dir.
How do I deal with the Windows executable and the MacOS dmg?

**Comment by giordano** [comment]
I think 7zip should be able to unpack a dmg

**Comment by ViralBShah** [comment]
I believe the right thing to do here is to make the Julia buildbots upload tarballs for all platforms. 

In the meanwhile I do believe you can find Linux tools to extract files from Windows installer executables.

**Comment by ViralBShah** [comment]
I believe the right thing to do here is to make the Julia buildbots upload tarballs for all platforms. 

In the meanwhile I do believe you can find Linux tools to extract files from Windows installer executables.

File: J/Julia/build_tarballs.jl [hunk 2]
```diff
@@ -3,99 +3,68 @@
...
-products(prefix) = Product[
-    ExecutableProduct(prefix, "julia", :julia),
-    LibraryProduct(prefix, "libjulia", :libjulia),
+products = Product[
+    ExecutableProduct("julia", :julia),
+    LibraryProduct("libjulia", :libjulia),
 ]
 
 # Dependencies that must be installed before this package can be built
 dependencies = [
```
Comment (giordano) [inline J/Julia/build_tarballs.jl]: Doesn't this still need binary dependencies?

File: J/Julia/build_tarballs.jl (hunk 1 above)
Comment (ViralBShah) [reply J/Julia/build_tarballs.jl]: We don't build Julia for all the supported platforms upstream (such as musl).

File: J/Julia/build_tarballs.jl (hunk 2 above)
Comment (ViralBShah) [reply J/Julia/build_tarballs.jl]: I don't think so. But maybe I am forgetting something.

**Comment by jstrube** [comment]
One important aspect of this is that it would have to package the includes. Am I right, @barche?
Not sure how a jll would accomplish that... Thoughts?

**Comment by barche** [comment]
@ViralBShah  I second the idea of using the official binaries, but extracting the official binaries on Linux failed for me for mac and/or Windows when I tried it (around the release of Julia 1.0, I think I tried with 7z).

@jstrube Yes, the includes are needed, but they are there in the official distribution and should also be in the jll tarball?

**Comment by ViralBShah** [comment]
I wonder what the CI systems do on mac and windows.

**Comment by giordano** [comment]
> I wonder what the CI systems do on mac and windows to get the official Julia binaries.

They have native systems, we don't

**Comment by ViralBShah** [comment]
> > I wonder what the CI systems do on mac and windows to get the official Julia binaries.
> 
> They have native systems, we don't

I thought there may be some intermediate tarballs from the buildbots that they might be using. But that is not the case - https://github.com/JuliaCI/install-julia/blob/master/install-julia.sh

**Comment by staticfloat** [comment]
It looks like our `.dmg` files are very unfriendly to extraction with `7z`.  You can `apk add p7zip` to get it in the build environment.

* To extract from `.exe` files, you do `7z x <path to .exe>`.  That will extract another `.exe`, which you do `7z x <path to new .exe>` on again.  That will result in the `bin`, `etc`, `include`, etc... directories plopped into the current directory.

* To extract from `.dmg` files, you can do the same, but one of the hfs filesystems it tries to extract is HUGE uncompressed (I stopped it after it tried to extract 50GB of zeros) so we would need to use a different tool here, I think.

Viral is probably right in that the best thing for us to do is to generate tarballs on the buildbots.

**Comment by jstrube** [comment]
I don't understand where this build error comes from: "can't dlopen library" https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=1283&view=logs&j=ed5bc0cf-9e11-56fd-407f-0f551f71f6db&t=afded763-3b68-5ccd-6b86-d980edfd7e43&l=1893
Especially on x86_64-linux-gnu this shouldn't happen... At least that one worked for me locally...

**Comment by ViralBShah** [comment]
Ok, who can help with making this happen?

**Comment by barche** [comment]
> Ok, who can help with making this happen?

The problem is not that it's hard or much work, but to repackage the official Julia release as a JLL, we need .tar.gz files for all platforms, and MacOS and Windows are missing (available as dmg and .exe instead). So it needs intervention from someone who can make the official binary build system generate these. Ideally, they should also use a naming scheme compatible with the `$target` variable in BB, but that's optional since with judicious use of the shell `case` statement it's easy enough to work around.

**Comment by ViralBShah** [comment]
@staticfloat Do you know who in the JuliaCI / buildbot area can help upload tarballs of these binaries going forward? For now, should we just do it manually?

**Comment by staticfloat** [comment]
We've made some progress on `master` recently, @musm has been helping to unify the `make binary-dist` targets; we should see if we can get the buildbots to upload not only the `.exe` and `.dmg` forms of Julia, but also the `.tar.gz`.  @musm sound like something fun to do?  I'd be happy to walk you through the necessary modifications to https://github.com/JuliaCI/julia-buildbot/blob/master/master/package.py

**Comment by ViralBShah** [comment]
Perhaps @DilumAluthge can help?

**Comment by giordano** [comment]
Superseded by #1048

**Comment by staticfloat** [comment]
https://github.com/JuliaBinaryWrappers/Julia_jll.jl

**Comment by ViralBShah** [comment]
It would perhaps be better to start using these in CI and such. @dilumaluthge

**Comment by ViralBShah** [comment]
@staticfloat Do these get uploaded automatically when we build a new release. Not seeing 1.4.2. 

@ararslan

**Comment by staticfloat** [comment]
No, someone has to update the builder in Yggdrasil and push a build through.

**Comment by ViralBShah** [comment]
I meant the upstream ArchiveSources - those are the ones that are failing.

**Comment by ViralBShah** [comment]
Opened https://github.com/JuliaPackaging/Yggdrasil/pull/1084 to figure out how to get Julia 1.4.2 in.

**Pull request closed without merging**